src/stim/search/sat/wcnf.cc

```diff
@@ -0,0 +1,229 @@
+      // WDIMACS clause format: weight var1 var2 ...
+      // To show negation of a variable, the index should be negated.
+      ss << quantized_weight(num_distinct_weights, top, clause.weight);
+      for (size_t i=0; i<clause.vars.size(); ++i) {
+        BoolRef var = clause.vars[i];
+        // Variables are 1-indexed
+        if (var.negated) {
+          ss << " -" << (var.variable + 1);
+        } else {
+          ss << " " << (var.variable + 1);
+        }
+      }
+      // Each clause ends with 0
+      ss << " 0\n";
+    }
+    return ss.str();
+  }
+};
+
+std::string stim::shortest_undetectable_logical_error_wcnf(const DetectorErrorModel &model, size_t num_distinct_weights) {
+  MaxSATInstance inst;
+
+  size_t num_observables = model.count_observables();
+  size_t num_detectors = model.count_detectors();
+  size_t num_errors = model.count_errors();
+  if (num_observables == 0 or num_detectors == 0 or num_errors == 0) {
+    std::stringstream err_msg;
+    err_msg << "Failed to find any logical errors.";
+    if (num_observables == 0) {
+        err_msg << "\n    WARNING: NO OBSERVABLES. The circuit or detector error model didn't define any observables, "
+                   "making it vacuously impossible to find a logical error.";
+    }
+    if (num_detectors == 0) {
+        err_msg << "\n    WARNING: NO DETECTORS. The circuit or detector error model didn't define any detectors.";
+    }
+    if (num_errors == 0) {
+        err_msg << "\n    WARNING: NO ERRORS. The circuit or detector error model didn't include any errors, making it "
+                   "vacuously impossible to find a logical error.";
+    }
+    throw std::invalid_argument(err_msg.str());
+  }
+  if (num_distinct_weights == 0) {
+    throw std::invalid_argument("num_distinct_weights must be >= 1");
+  }
+
+  MaxSATInstance instance;
+  // Create a boolean variable for each error, which indicates whether it is activated.
+  std::vector<BoolRef> errors_activated;
+  for (size_t i=0; i<num_errors; ++i) {
+    errors_activated.push_back(instance.new_bool());
+  }
+
+  std::vector<BoolRef> detectors_activated(num_detectors, BoolRef::False());
+  std::vector<BoolRef> observables_flipped(num_observables, BoolRef::False());
+
+  size_t error_index = 0;
+  model.iter_flatten_error_instructions([&](const DemInstruction &e) {
+      if (e.arg_data[0] != 0) {
+        BoolRef err_x = errors_activated[error_index];
+        // Add parity contribution to the detectors and observables
+        for (const auto &t : e.target_data) {
+            if (t.is_relative_detector_id()) {
+              detectors_activated[t.val()] = instance.Xor(detectors_activated[t.val()], err_x);
+            } else if (t.is_observable_id()) {
+              observables_flipped[t.val()] = instance.Xor(observables_flipped[t.val()], err_x);
+            }
+        }
+        // Add a soft clause for this error to be inactive
+        Clause clause;
+        clause.add_var(~err_x);
```

Comment:
Since p>0.5 will lead to a disallowed negative weight, it might be worth handling the p>0.5 case with `clause.add_var(err_x);` and then flipping the sign of the weight?

Reply:
On second thoughts it may be best to leave it as is, since by default (unweighted case) the method finds the _shortest_ logical error, not the most probable.

Reply:
Hmm, good catch!
I agree with both your first and second thoughts and I am troubled by the potential for accidental mistakes in using this.

It's unclear to me what the best choice of API is that balances utility with safety.

Do you think it might be appropriate to throw an InvalidArgument or cerr a warning if there is any error mechanism with probability >= 0.5? This way it's at least a visible error. WDYT?

Reply:
Yeah agree that just throwing an InvalidArgument about requiring p<0.5 is probably safest (and tbf it actually already does this with the message `"Clauses must have positive weight or HARD_CLAUSE_WEIGHT."` from line 58)

Reply:
discussed offline and added another argument `weighted=False` where if it's True, does the p > 0.5 accounting Oscar suggested above.

Reply:
The docstring should probably explicitly say how p>0.5 errors are inverted if `weight_bins > 1`.

Reply:
Done